### PR TITLE
paralleled-tx : group

### PR DESCRIPTION
**Pull request by scf0220**
- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: app/app_parallel.go [hunk 1]
```diff
@@ -20,10 +21,12 @@ func updateFeeCollectorHandler(bk bank.Keeper, sk supply.Keeper) sdk.UpdateFeeCo
 
 // evmTxFeeHandler get tx fee for evm tx
 func evmTxFeeHandler() sdk.GetTxFeeHandler {
-	return func(ctx sdk.Context, tx sdk.Tx) (fee sdk.Coins, isEvm bool) {
+	return func(ctx sdk.Context, tx sdk.Tx) (fee sdk.Coins, isEvm bool, from string, to *ethcmn.Address) {
 		if evmTx, ok := tx.(*evmtypes.MsgEthereumTx); ok {
 			isEvm = true
 			_ = evmTx.VerifySig(evmTx.ChainID(), ctx.BlockHeight())
```
Comment (scf0220) [inline app/app_parallel.go]: check if need err

File: libs/cosmos-sdk/baseapp/abci.go [hunk 2]
```diff
@@ -175,6 +175,11 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
 		res = app.endBlocker(app.deliverState.ctx, req)
 	}
 
+	go func() {
+		if app.deliverState != nil && app.deliverState.ms != nil {
+			app.deliverState.ms.Write()
```
Comment (scf0220) [inline libs/cosmos-sdk/baseapp/abci.go]: delete it and issue

File: app/app_parallel.go (hunk 1 above)
Comment (scf0220) [reply app/app_parallel.go]: no need to check err , only use from for group

File: libs/cosmos-sdk/baseapp/abci.go [hunk 3]
```diff
@@ -241,6 +241,7 @@ func (app *BaseApp) addCommitTraceInfo() {
 // against that height and gracefully halt if it matches the latest committed
 // height.
 func (app *BaseApp) Commit(req abci.RequestCommit) abci.ResponseCommit {
+
```
Comment (MinZhu09) [inline libs/cosmos-sdk/baseapp/abci.go:262]: delete

File: libs/cosmos-sdk/baseapp/baseapp_parallel.go [hunk 4]
```diff
@@ -159,101 +179,156 @@ func (app *BaseApp) runTxs(txs [][]byte) []*abci.ResponseDeliverTx {
...
+			txRs := res.resp
 			deliverTxs[txIndex] = &txRs
-			res.Collect(asCache)
-			res.Commit()
-			app.fixFeeCollector(app.parallelTxManage.indexMapBytes[txIndex])
-			if !s.reRun {
-				app.deliverState.ctx.BlockGasMeter().ConsumeGas(sdk.Gas(res.resp.GasUsed), "unexpected error")
-			}
+
+			app.deliverState.ctx.BlockGasMeter().ConsumeGas(sdk.Gas(res.resp.GasUsed), "unexpected error")
```
Comment (MinZhu09) [inline libs/cosmos-sdk/baseapp/baseapp_parallel.go:255]: don't take care of the case of ErrorGasOverflow

File: libs/cosmos-sdk/types/context.go [hunk 5]
```diff
@@ -212,11 +219,6 @@ func (c *Context) SetFrom(from string) *Context {
 	return c
 }
 
-func (c *Context) SetAsync(isAsync bool) *Context {
-	c.isAsync = isAsync
```
Comment (MinZhu09) [inline libs/cosmos-sdk/types/context.go:216]: should delete this variable in Context

File: libs/cosmos-sdk/baseapp/baseapp_parallel.go [hunk 6]
```diff
@@ -159,101 +179,156 @@ func (app *BaseApp) runTxs(txs [][]byte) []*abci.ResponseDeliverTx {
...
 
-			if res.Conflict(asCache) || overFlow(currentGas, res.resp.GasUsed, maxGas) {
+			if pm.newIsConflict(res) || overFlow(currentGas, res.resp.GasUsed, maxGas) {
 				rerunIdx++
-				s.reRun = true
-				res = app.deliverTxWithCache(txs[txIndex])
+
+				// conflict rerun tx
+				if pm.extraTxsInfo[txIndex].isEvm {
+					app.fixFeeCollector()
```
Comment (MinZhu09) [inline libs/cosmos-sdk/baseapp/baseapp_parallel.go:234]: should call fixFeeCollector() when the tx is non-evm tx

File: libs/cosmos-sdk/baseapp/baseapp_parallel.go [hunk 7]
```diff
@@ -264,281 +339,300 @@ func (app *BaseApp) deliverTxWithCache(txByte []byte) *executeResult {
...
-	fmt.Println("BestBlock", l.bestBlock.string(), "Concurrency Rate", 1-float64(l.bestBlock.reRunTxs)/float64(l.bestBlock.txs))
-	fmt.Println("TerribleBlock", l.terribleBlock.string(), "Concurrency Rate", 1-float64(l.terribleBlock.reRunTxs)/float64(l.terribleBlock.txs))
+func (f *parallelTxManager) SetCurrentIndex(txIndex int, res *executeResult) {
+	if res.ms == nil {
+		return
+	}
+
+	f.mu.Lock()
+	res.ms.IteratorCache(true, func(key string, value []byte, isDirty bool, isdelete bool, storeKey sdk.StoreKey) bool {
+		if isDirty {
```
Comment (MinZhu09) [inline libs/cosmos-sdk/baseapp/baseapp_parallel.go]: there is no need to check isDirty

**Review by MinZhu09 — DISMISSED** [review]

**Review by xiangjianmeng — APPROVED** [review]

**Pull request merged**